### PR TITLE
[ASDisplayNode] Create placeholders off main when possible.

### PR DESCRIPTION
**Pull request by appleguy**
This is a fix for https://github.com/facebook/AsyncDisplayKit/issues/2620,
which is a followup to https://github.com/facebook/AsyncDisplayKit/pull/2610

This will ensure thread affinity is respected for .contents but still run
placeholderImage off-main when possible (node isn't loaded).

### CODE REVIEW

File: AsyncDisplayKit/ASDisplayNode.mm [hunk 1]
```diff
@@ -1306,21 +1306,27 @@ - (void)_pendingLayoutTransitionDidComplete
...
       return;
     }
-
-    if (!_placeholderImage) {
+    
+    // If we've displayed our contents, we don't need a placeholder.
+    // Contents is a thread-affined property and can't be read off main after loading.
+    if (self.isNodeLoaded) {
       ASPerformBlockOnMainThread(^{
         if (self.contents == nil) {
```
Comment (maicki) [inline AsyncDisplayKit/ASDisplayNode.mm]: Can we move the `self.contents` block higher or do we have to check it in here after running on main? Something like:

```Objc
if (self.contents == nil) {
  if (self.isNodeLoaded) {
    ASPerformBlockOnMainThread(_placeholderImage = [self placeholderImage];)
  } else {
   _placeholderImage = [self placeholderImage];
  }
}
```

File: AsyncDisplayKit/ASDisplayNode.mm [hunk 2]
```diff
@@ -1306,21 +1306,27 @@ - (void)_pendingLayoutTransitionDidComplete
...
   // This is also because measurement is usually asynchronous, but placeholders need to be set up synchronously.
   // First measurement is guaranteed to be before the node is onscreen, so we can create the image async. but still have it appear sync.
-  if (_placeholderEnabled && [self _displaysAsynchronously]) {
+  if (_placeholderEnabled && !_placeholderImage && [self _displaysAsynchronously]) {
     
     // Zero-sized nodes do not require a placeholder.
     ASLayout *layout = _calculatedDisplayNodeLayout->layout;
     CGSize layoutSize = (layout ? layout.size : CGSizeZero);
-    if (CGSizeEqualToSize(layoutSize, CGSizeZero)) {
+    if (layoutSize.width * layoutSize.height <= 0.0) {
```
Comment (maicki) [inline AsyncDisplayKit/ASDisplayNode.mm]: Can you addd a `()` around `layoutSize.width * layoutSize.height` just to make it a bit more clear

File: AsyncDisplayKit/ASDisplayNode.mm (hunk 1 above)
Comment (Adlai-Holler) [inline AsyncDisplayKit/ASDisplayNode.mm]: @maicki We can't read `contents` off the main thread =/

**Comment by appleguy** [comment]
In a bit of a rush for Thanksgiving -- I agree it would be good to add parentheses when we have a moment to do so.

**Pull request merged**